gtsfm/averaging/rotation/shonan.py

```diff
@@ -56,23 +69,58 @@ def run(self, num_images: int, i2Ri1_dict: Dict[Tuple[int, int], Optional[Rot3]]
 
         for (i1, i2), i2Ri1 in i2Ri1_dict.items():
             if i2Ri1 is not None:
-                between_factors.append(
-                    BetweenFactorPose3(
-                        i2,
-                        i1,
-                        Pose3(
-                            i2Ri1,
-                            np.zeros(
-                                3,
-                            ),
-                        ),
-                        noise_model,
-                    )
-                )
+                # ignore translation during rotation averaging
+                i2Ti1 = Pose3(i2Ri1, np.zeros(3))
+                between_factors.append(BetweenFactorPose3(i2, i1, i2Ti1, noise_model))
 
         obj = ShonanAveraging3(between_factors, shonan_params)
 
         initial = obj.initializeRandomly()
         result_values, _ = obj.run(initial, self._p_min, self._p_max)
 
-        return [result_values.atRot3(i) for i in range(num_images)]
+        return [result_values.atRot3(i) if result_values.exists(i) else None for i in range(num_connected_nodes)]
+
+    def run(self, num_images: int, i2Ri1_dict: Dict[Tuple[int, int], Optional[Rot3]]) -> List[Optional[Rot3]]:
+        """Wrapper that re-orders keys to prepare a connected graph, w/ N keys ordered [0,...,N-1].
+
+        Args:
+            num_images: number of poses.
+            i2Ri1_dict: relative rotations for each image pair-edge as dictionary (i1, i2): i2Ri1.
+
+        Returns:
+            Global rotations for each camera pose, i.e. wRi, as a list. The number of entries in the list is
+                `num_images`. The list may contain `None` where the global rotation could not be computed (either
+                underconstrained system or ill-constrained system), or where the camera pose had no valid observation
+                in the input to run().
+        """
+        connected_nodes = set()
+        for (i1, i2) in i2Ri1_dict.keys():
+            connected_nodes.add(i1)
+            connected_nodes.add(i2)
+
+        connected_nodes = sorted(list(connected_nodes))
+
+        # given original index, get back new temporary index
+        reordered_idx_map = {}
+        for i in range(num_images):
+            if i in connected_nodes:
+                reordered_idx_map[i] = connected_nodes.index(i)
+
+        # now, re-order all of the indices
+        i2Ri1_dict_reordered = {}
+        for (i1, i2), i2Ri1 in i2Ri1_dict.items():
+            i1_ = reordered_idx_map[i1]
+            i2_ = reordered_idx_map[i2]
+            i2Ri1_dict_reordered[(i1_, i2_)] = i2Ri1
+
+        wRi_list_subset = self.run_w_consecutive_ordering(
+            num_connected_nodes=len(connected_nodes), i2Ri1_dict=i2Ri1_dict_reordered
+        )
+
+        wRi_list = [None] * num_images
+        for i in range(num_images):
+            if i in connected_nodes:
```

Comment:
please check `if i in reordered_idx_map` instead. Since that is the one being used below. 
Also we might not be losing much readability by `wRi_list[i] = wRi_list_subset[reordered_idx_map[i]]`, but thats upto you.

Reply:
We can directly loop over the `connected_nodes`?

```
for remapped_i, original_i in enumerate(connected_nodes):
    wRi_list[original_i] = wRi_list_subset[remapped_i]
```